File: HW10/task1.cpp
#include <iostream>
#include <vector>

void dfs(int v, std::vector<int> &used, std::vector<std::vector<int>> &graph, int &cnt) {
	used[v] = cnt;
	for (auto u : graph[v]) {
		if (!used[u]) {
			dfs(u, used, graph, cnt);
		}
	}
}

void find_comps(std::vector<int> &used, std::vector<std::vector<int>> &graph) {
	int cnt;
	cnt = 0;
	for (int i = 0; i < used.size(); ++i)
		if (!used[i]) {
			cnt++;
			dfs(i, used, graph, cnt);
		}
	std::cout << cnt << "\n";
	for (int i : used) {
		std::cout << i << " ";
	}
}

int main() {
	int n, m;
	int v, u;
	std::cin >> n >> m;
	std::vector<int> used(n, 0);
	std::vector<std::vector<int>> graph(n);
	while (m--) {
		std::cin >> v >> u;
		v--;
		u--;
		graph[v].push_back(u);
		graph[u].push_back(v);
	}
	find_comps(used, graph);
	return 0;
}

File: HW10/task1.py
from sys import setrecursionlimit
import threading
import sys


def dfs(v, visited, graph, cnt):
    visited[v] = cnt
    for w in graph[v]:
        if visited[w] == False:  # посещён ли текущий сосед?
            dfs(w, visited, graph, cnt)

def main():
    n, m = map(int, sys.stdin.readline().split())
    graph = [[] for _ in range(n)]
    visited = [False] * n
    while m:
        v, u = map(int, sys.stdin.readline().split())
        graph[v - 1].append(u - 1)
        m -= 1
    cnt = 0
    for v in range(n):
        if visited[v] == False:
            cnt += 1
            dfs(v, visited, graph, cnt)
    print(*visited)


if __name__ == "__main__":
    main()

setrecursionlimit(10 ** 9)
threading.stack_size(2 ** 26)  # лучше использовать именно эту константу
thread = threading.Thread(target=main)
thread.start()



